cassandra/src/main/java/io/jaegertracing/spark/dependencies/cassandra/CassandraDependenciesJob.java

```diff
@@ -0,0 +1,210 @@
+    public Builder jars(String... jars) {
+      this.jars = jars;
+      return this;
+    }
+
+    /** Keyspace to store dependency rowsToLinks. Defaults to "zipkin" */
+    public Builder keyspace(String keyspace) {
+      checkNoNull("keyspace", keyspace);
+      this.keyspace = keyspace;
+      return this;
+    }
+
+    /** Day (in epoch milliseconds) to process dependencies for. Defaults to today. */
+    public Builder day(long day) {
+      this.day = midnightUTC(day);
+      return this;
+    }
+
+    /** Ensures that logging is setup. Particularly important when in cluster mode. */
+    public Builder logInitializer(Runnable logInitializer) {
+      checkNoNull("logInitializer", logInitializer);
+      this.logInitializer = logInitializer;
+      return this;
+    }
+
+    /** Comma separated list of hosts / IPs part of Cassandra cluster. Defaults to localhost */
+    public Builder contactPoints( String contactPoints) {
+      this.contactPoints = contactPoints;
+      return this;
+    }
+
+    /** The local DC to connect to (other nodes will be ignored) */
+    public Builder localDc(String localDc) {
+      this.localDc = localDc;
+      return this;
+    }
+
+    public CassandraDependenciesJob build() {
+      return new CassandraDependenciesJob(this);
+    }
+  }
+
+  private static String getEnv(String key, String defaultValue) {
+    String result = System.getenv(key);
+    return result != null ? result : defaultValue;
+  }
+
+  final String keyspace;
+  final long day;
+  final String dateStamp;
+  final SparkConf conf;
+  final Runnable logInitializer;
+
+  CassandraDependenciesJob(Builder builder) {
+    this.keyspace = builder.keyspace;
+    this.day = builder.day;
+    SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd");
+    df.setTimeZone(TimeZone.getTimeZone("UTC"));
+    this.dateStamp = df.format(new Date(builder.day));
+    this.conf = new SparkConf(true)
+        .setMaster(builder.sparkMaster)
+        .setAppName(getClass().getName());
+    conf.set("spark.cassandra.connection.host", parseHosts(builder.contactPoints));
+    conf.set("spark.cassandra.connection.port", parsePort(builder.contactPoints));
+    if (builder.localDc != null) conf.set("connection.local_dc", builder.localDc);
+    if (builder.jars != null) conf.setJars(builder.jars);
+    for (Map.Entry<String, String> entry : builder.sparkProperties.entrySet()) {
+      conf.set(entry.getKey(), entry.getValue());
+    }
+    this.logInitializer = builder.logInitializer;
+  }
+
+  public void run() {
+    long microsLower = day * 1000;
+    long microsUpper = (day * 1000) + TimeUnit.DAYS.toMicros(1) - 1;
+
+    log.info("Running Dependencies job for {}: {} ≤ Span.timestamp {}", dateStamp, microsLower,
+        microsUpper);
+
+    JavaSparkContext sc = new JavaSparkContext(conf);
+    List<Dependency> dependencies = javaFunctions(sc).cassandraTable(keyspace, "traces", mapRowTo(Span.class))
```

Comment:
@Jiri-Kremser could you please have a look at these lines?

Reply:
> you please have a look at these lines?

nice lines :)


perhaps

```java
   .flatMapValues(foo)
   .values()
   .mapToPair(bar)
```

could be all expressed in the first `flatMapValues`. 

RDDs are fun to write, but can be slow sometimes, if you want optimization for free, consider using the DataFrame abstraction in [should be](https://github.com/datastax/spark-cassandra-connector/blob/master/doc/14_data_frames.md) in the C*-spark connector.